File: backend/app/models/capture.py
"""Database models for captures and drafts."""
from sqlalchemy import Column, String, Integer, Float, DateTime, Enum, Text, Boolean
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.sql import func
from datetime import datetime
import enum

from app.db.base import Base


class CaptureType(str, enum.Enum):
    """Type of user capture input."""
    VOICE = "voice"
    TEXT = "text"
    IMAGE = "image"


class CaptureState(str, enum.Enum):
    """State of capture processing."""
    QUEUED = "queued"
    PROCESSING = "processing"
    PARSED = "parsed"
    CONFIRMED = "confirmed"
    SYNCING = "syncing"
    SYNCED = "synced"
    ERROR = "error"
    NEEDS_REVIEW = "needs_review"


class DraftType(str, enum.Enum):
    """Type of draft item."""
    TASK = "task"
    EVENT = "event"
    NOTE = "note"


class Capture(Base):
    """User capture input (voice, text, or image)."""
    __tablename__ = "captures"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(String, nullable=False, index=True)
    capture_type = Column(Enum(CaptureType), nullable=False)
    state = Column(Enum(CaptureState), default=CaptureState.QUEUED, nullable=False)
    
    # Raw input data
    raw_text = Column(Text, nullable=True)  # For text capture or transcribed voice
    audio_url = Column(String, nullable=True)  # Cloud storage URL for audio
    image_url = Column(String, nullable=True)  # Cloud storage URL for image
    
    # Processing metadata
    ai_confidence = Column(Float, nullable=True)
    error_message = Column(Text, nullable=True)
    retry_count = Column(Integer, default=0)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    processed_at = Column(DateTime(timezone=True), nullable=True)


class Draft(Base):
    """AI-parsed structured draft (task or event)."""
    __tablename__ = "drafts"

    id = Column(Integer, primary_key=True, index=True)
    capture_id = Column(Integer, nullable=False, index=True)
    user_id = Column(String, nullable=False, index=True)
    
    # Classification
    draft_type = Column(Enum(DraftType), nullable=False)
    ai_confidence = Column(Float, nullable=False)
    
    # Extracted fields
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    due_date = Column(DateTime(timezone=True), nullable=True)
    priority = Column(String, nullable=True)  # high, medium, low
    location = Column(String, nullable=True)
    
    # Event-specific fields
    start_time = Column(DateTime(timezone=True), nullable=True)
    end_time = Column(DateTime(timezone=True), nullable=True)
    
    # Sync state
    is_confirmed = Column(Boolean, default=False)
    google_task_id = Column(String, nullable=True)
    google_event_id = Column(String, nullable=True)
    sync_state = Column(Enum(CaptureState), default=CaptureState.PARSED)
    
    # Metadata
    extracted_entities = Column(JSONB, nullable=True)  # Full NLP output
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    synced_at = Column(DateTime(timezone=True), nullable=True)

File: mobile/lib/models/capture.dart
// Data model for Capture

enum CaptureType { voice, text, image }

enum CaptureState {
  queued,
  processing,
  parsed,
  confirmed,
  syncing,
  synced,
  error,
  needsReview
}

class Capture {
  final int? id;
  final String userId;
  final CaptureType captureType;
  final CaptureState state;
  final String? rawText;
  final String? audioUrl;
  final String? imageUrl;
  final String? audioData;  // Base64 encoded audio for upload
  final String? imageData;  // Base64 encoded image for upload
  final double? aiConfidence;
  final String? errorMessage;
  final int retryCount;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final DateTime? processedAt;

  Capture({
    this.id,
    required this.userId,
    required this.captureType,
    required this.state,
    this.rawText,
    this.audioUrl,
    this.imageUrl,
    this.audioData,
    this.imageData,
    this.aiConfidence,
    this.errorMessage,
    this.retryCount = 0,
    required this.createdAt,
    this.updatedAt,
    this.processedAt,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'capture_type': captureType.name,
      'state': state.name,
      'raw_text': rawText,
      'audio_url': audioUrl,
      'image_url': imageUrl,
      'audio_data': audioData,
      'image_data': imageData,
      'ai_confidence': aiConfidence,
      'error_message': errorMessage,
      'retry_count': retryCount,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
      'processed_at': processedAt?.toIso8601String(),
    };
  }

  factory Capture.fromJson(Map<String, dynamic> json) {
    return Capture(
      id: json['id'],
      userId: json['user_id'],
      captureType: CaptureType.values.firstWhere(
        (e) => e.name == json['capture_type'],
      ),
      state: CaptureState.values.firstWhere(
        (e) => e.name == json['state'],
      ),
      rawText: json['raw_text'],
      audioUrl: json['audio_url'],
      imageUrl: json['image_url'],
      audioData: json['audio_data'],
      imageData: json['image_data'],
      aiConfidence: json['ai_confidence']?.toDouble(),
      errorMessage: json['error_message'],
      retryCount: json['retry_count'] ?? 0,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
      processedAt: json['processed_at'] != null
          ? DateTime.parse(json['processed_at'])
          : null,
    );
  }

  Map<String, dynamic> toSqlite() {
    return {
      'id': id,
      'user_id': userId,
      'capture_type': captureType.name,
      'state': state.name,
      'raw_text': rawText,
      'audio_url': audioUrl,
      'image_url': imageUrl,
      'ai_confidence': aiConfidence,
      'error_message': errorMessage,
      'retry_count': retryCount,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
      'processed_at': processedAt?.toIso8601String(),
    };
  }

  factory Capture.fromSqlite(Map<String, dynamic> map) {
    return Capture.fromJson(map);
  }
}
